File: 501-600/551-560/552-studentAttendanceRecord2/studentAttendanceRecord2.cpp
// 矩阵快速幂，answer from discussion

class Solution {
public:
    static constexpr int MOD = 1'000'000'007;

    vector<vector<long>> pow(vector<vector<long>> mat, int n) {
        vector<vector<long>> ret = {{1, 0, 0, 0, 0, 0}};
        while (n > 0) {
            if ((n & 1) == 1) {
                ret = multiply(ret, mat);
            }
            n >>= 1;
            mat = multiply(mat, mat);
        }
        return ret;
    }

    vector<vector<long>> multiply(vector<vector<long>> a, vector<vector<long>> b) {
        int rows = a.size(), columns = b[0].size(), temp = b.size();
        vector<vector<long>> c(rows, vector<long>(columns));
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < columns; j++) {
                for (int k = 0; k < temp; k++) {
                    c[i][j] += a[i][k] * b[k][j];
                    c[i][j] %= MOD;
                }
            }
        }
        return c;
    }

    int checkRecord(int n) {
        vector<vector<long>> mat = {{1, 1, 0, 1, 0, 0}, {1, 0, 1, 1, 0, 0}, {1, 0, 0, 1, 0, 0}, {0, 0, 0, 1, 1, 0}, {0, 0, 0, 1, 0, 1}, {0, 0, 0, 1, 0, 0}};
        vector<vector<long>> res = pow(mat, n);
        long sum = accumulate(res[0].begin(), res[0].end(), 0ll);
        return (int)(sum % MOD);
    }
};

// 作者：LeetCode-Solution
// 链接：https://leetcode-cn.com/problems/student-attendance-record-ii/solution/xue-sheng-chu-qin-ji-lu-ii-by-leetcode-s-kdlm/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

File: 501-600/551-560/552-studentAttendanceRecord2/studentAttendanceRecord2.py
class Solution(object):
    def checkRecord(self, n):
        """
        :type n: int
        :rtype: int
        """

        dp = [[0] * 3 for _ in range(2)]
        dp[0][0] = 1
        MOD = 10**9 + 7

        for _ in range(n):
            new_dp = [[0] * 3 for _ in range(2)]
            for a in range(2):
                for l in range(3):
                    if a < 1:
                        new_dp[a+1][0] = (new_dp[a+1][0] + dp[a][l]) % MOD
                    if l < 2:
                        new_dp[a][l+1] = (new_dp[a][l+1] + dp[a][l]) % MOD

                    new_dp[a][0] = (new_dp[a][0] + dp[a][l]) % MOD

            dp = new_dp

        return sum(map(sum, dp)) % MOD